### PR TITLE
Add key-value-storage to the InterfaceInfo

### PR DESCRIPTION
**Pull request by mamueluth**
This PR adds the possibility to define custom parameters (key-value-storage) per Interface (Commad-/StateInterface). This is useful to define a .ros2_control.xacro for a hardware driver that uses for example Modbus for communication. In this case, you want to be able to define e.g. which register is used to retrieve/write the data for each Command-/StateInteface Individually.

Has been tested and works. Test have been added. 

The exception as been altered as I did some mistake while defining the URDF and the thrown exception "invalid URDF passed in to robot parser" is not very useful at all. I would prefer this to be merged  as it now gives some useful hint what was gone wrong. If i should move  the two lines of code to an other PR as its not really related please let me know.

### CODE REVIEW

**Review by christophfroehlich — CHANGES_REQUESTED** [review]
This looks valuable and the implementation is straight forward, thanks!

Just one usual comment from my side, please add this to the docs:

- [ ] maybe [here](https://control.ros.org/master/doc/ros2_control/hardware_interface/doc/hardware_interface_types_userdoc.html) or create a new page.
- [ ] into the [release notes](https://github.com/ros-controls/ros2_control/pull/1477/files#diff-0c007bfed7d9684b385392a2b58811b75b952bb7702b2571a6f84d2e0cd9e602). If it get not merged soon, just comment here and I'll add that later.

**Comment by christophfroehlich** [comment]
@mamueluth can you please rebase this on master to run the checks? (we can't do that because this PR comes from an organization)

**Comment by christophfroehlich** [comment]
Fyi, CI looks good but codecov needs https://github.com/ros-controls/ros2_control/pull/1504 and the other failures are already reported  https://github.com/ros-controls/ros2_control/issues/1367 and not related imho.

**Comment by mamueluth** [comment]
@christophfroehlich  I added a overview to the docs. There the key-value storage is described in a generic example. We could explain the other xml tags there as well.
Release notes let me know if i should add.

File: hardware_interface/doc/hardware_interface_types_userdoc.rst [hunk 1]
```diff
@@ -9,6 +9,38 @@ The ``ros2_control`` framework provides a set of hardware interface types that c
...
+
+    <ros2_control name="Name_of_the_hardware" type="system|sensor|actuator">
+      <hardware>
+        <plugin>library_name/ClassName</plugin>
+        <!-- added to hardware_parameters -->
+        <param name="example_param">value</param>
+      </hardware>
+      <joint|gpio|sensor name="name_of_the_component">
+        <command_interface name="interface_name">
+          <!-- All of them are optional. Data_type and size are used for GPIOs. Size length of array. -->
```
Comment (destogl) [inline hardware_interface/doc/hardware_interface_types_userdoc.rst:29]:
```suggestion
          <!-- All of them are optional. `data_type` and `size` are used for GPIOs. Size is length of an array. -->
```

**Review by destogl — APPROVED** [review]
Very nice, thanks!

**Review by destogl — APPROVED** [review]

**Review by destogl — DISMISSED** [review]

File: hardware_interface/doc/hardware_interface_types_userdoc.rst [hunk 2]
```diff
@@ -9,6 +9,38 @@ The ``ros2_control`` framework provides a set of hardware interface types that c
 a hardware component for a specific robot or device.
 The following sections describe the different hardware interface types and their usage.
 
+Overview
+*****************************
+Hardware in ros2_control is described as URDF and internally parsed and encapsulated as ``HardwareInfo``.
+The definition can be found in the `ros2_control repository <https://github.com/ros-controls/ros2_control/blob/master/hardware_interface/include/hardware_interface/hardware_info.hpp>`_.
+You can check this classes to see what attributes are available for each of the xml tags.
```
Comment (saikishor) [inline hardware_interface/doc/hardware_interface_types_userdoc.rst:16]:
```suggestion
You can check this struct to see what attributes are available for each of the xml tags.
```

**Review by saikishor — APPROVED** [review]
Very nice!
Nice tests too!
Thanks for the nice work

File: hardware_interface/doc/hardware_interface_types_userdoc.rst (hunk 2 above)
Comment (mamueluth) [inline hardware_interface/doc/hardware_interface_types_userdoc.rst:16]:
```suggestion
You can check the structs defined there to see what attributes are available for each of the xml tags.
```

**Review by destogl — DISMISSED** [review]

File: hardware_interface/include/hardware_interface/hardware_info.hpp [hunk 3]
```diff
@@ -46,6 +46,10 @@ struct InterfaceInfo
   int size;
   /// (Optional) enable or disable the limits for the command interfaces
   bool enable_limits;
+  /// (Optional) Key-value pairs of command/stateInterface parameters. This is
+  /// useful for drivers that operate on protocols like modbus, where each
+  /// interface needs own address(register), datatype, ...
```
Comment (bmagyar) [inline hardware_interface/include/hardware_interface/hardware_info.hpp:51]:
```suggestion
  /// interface needs own address(register), datatype, etc.
```

**Review by destogl — APPROVED** [review]

**Review by bmagyar — APPROVED** [review]

**Review by christophfroehlich — CHANGES_REQUESTED** [review]
A short note in the release notes would be great 😇

**Review by christophfroehlich — APPROVED** [review]
(sorry for being late, we can add the release notes in a follow-up PR as well).

**Comment by mamueluth** [comment]
> A short note in the release notes would be great 😇

done

File: doc/release_notes.rst [hunk 4]
```diff
@@ -105,6 +105,10 @@ hardware_interface
 * Added ``get_hardware_info`` method to the hardware components interface to access the ``HardwareInfo`` instead of accessing the variable ``info_`` directly (`#1643 <https://github.com/ros-controls/ros2_control/pull/1643>`_)
 * With (`#1683 <https://github.com/ros-controls/ros2_control/pull/1683>`_) the ``rclcpp_lifecycle::State & get_state()`` and ``void set_state(const rclcpp_lifecycle::State & new_state)`` are replaced by ``rclcpp_lifecycle::State & get_lifecycle_state()`` and ``void set_lifecycle_state(const rclcpp_lifecycle::State & new_state)``. This change affects controllers and hardware. This is related to (`#1240 <https://github.com/ros-controls/ros2_control/pull/1240>`_) as variant support introduces ``get_state`` and ``set_state`` methods for setting/getting state of handles.
 
+InterfaceInfo
+*************
+* With (`#1421 <https://github.com/ros-controls/ros2_control/pull/1421>`_) a key-value storage is added to InterfaceInfo. This allows to define extra params with per Command-/StateInterface in the ``.ros2_control.xacro`` file.
+
```
Comment (saikishor) [inline doc/release_notes.rst:111]:
```suggestion
* With (`#1683 <https://github.com/ros-controls/ros2_control/pull/1683>`_) the ``rclcpp_lifecycle::State & get_state()`` and ``void set_state(const rclcpp_lifecycle::State & new_state)`` are replaced by ``rclcpp_lifecycle::State & get_lifecycle_state()`` and ``void set_lifecycle_state(const rclcpp_lifecycle::State & new_state)``. This change affects controllers and hardware. This is related to (`#1240 <https://github.com/ros-controls/ros2_control/pull/1240>`_) as variant support introduces ``get_state`` and ``set_state`` methods for setting/getting state of handles.
* With (`#1421 <https://github.com/ros-controls/ros2_control/pull/1421>`_) a key-value storage is added to InterfaceInfo. This allows to define extra params with per Command-/StateInterface in the ``.ros2_control.xacro`` file.

```

**Review by saikishor — CHANGES_REQUESTED** [review]
A minor detail

**Review by saikishor — APPROVED** [review]

**Pull request merged**